### PR TITLE
Update HPA to only account for available pods

### PR DESCRIPTION
**Pull request by LindseySaari**
## Summary

This PR updates the HPA configuration to account for **only** the available pods. Available pods are those that have passed the readiness probes and are actively serving traffic. The primary goal is to ensure that the autoscaling decisions are based on the actual capacity of the application by excluding pods that are not ready due to high backlogs when they fail the [readiness probe script](https://github.com/department-of-veterans-affairs/vsp-infra-application-manifests/pull/2956).

This was driven by the high backlog yesterday. We should not have a backlog this high ever and autoscaling should take care of this

Screenshot 2024-08-08 at 10 30 18 AM

## Related issue(s)
https://app.zenhub.com/workspaces/platform-product-team-633af4074573d06c3cda142a/issues/gh/department-of-veterans-affairs/va.gov-team/88350

## Testing done

This is the new HPA metric [shown as a visual](https://vagov.ddog-gov.com/metric/explorer?fromUser=false&start=1723084121429&end=1723127321429&paused=false#N4Ig7glgJg5gpgFxALlAGwIYE8D2BXJVEADxQEYAaELcqyKBAC1pEbghkcLIF8qo4AMwgA7CAgg4RKUAiwAHOChASAtnADOcAE4RNIKtrgBHPJoQaUAbVBGN8qVoD6gnNtUZCKiOq279VKY6epbINiAiGOrKQdpYZAYgUJ4YThr42gDGSsgg6gi6mZaBZnHKGABuMMjyeB4AdB7ETkxGGFAawHAiFchwANYaALTy2jhQPAAEAEZYk8D9eNNwTgLyaDhY6iIIPCB8oJHRubFYAEyJyQip6XhZOXmIhcUgp8oadciLy9oiiJppa4IOD1NYbLbdBD1IzrCCZDAaJyVDAQTDTNBwBZLFZHTTyDDZZAVRDDDDyCAUSbfFaZNB4DTA7ROXHIDoVDBDdkwHAVEZjKBDWn0xmTSkAQip2NWcHWm22CCJJKGZIgQw00AGEGMACoprN5tTpbKITs9gcIlEHqcAMyXFJpDLZZT5Z6JN65SrVWoNBw4NBOeH4zLiLBY5bMy32AlwRUWZXkiiGoUMnQR9SsjTszkYbm80bjQV0lPaPVzMMrMFyyFmiiHS0xUpYAAsduuDruTtyLrhL3dIA+qi+2N+-0RDM8IMrJqhMLQcIRTjwkXZqIw6MxhtxUcJxLjKspSaLjLTMbZHK5PL5BeTIvFkvDU-lsdJ5LVGv6Wt1MzLhsf1f2AC6VCuO4eCYKE4QYHOCLKAAshgxCTK0cDtC8IGqGBGANjo8T7LWIBQRAMG5AASjKc7whokwAIIrmiGKTIk6GYdhcQXOahHESAtEovRcCTAAKowbQdExbgYZgrFYLaHHQaEID0nAUAAPQIDg1xoEhwkoR0kwABSKPcOw5nAACUYmgZJuR6XppxkJM2qTKcZymZMQz6TaDlOY2LmucpHmNvZjnOa5jlkAADOFgFUAyWAYjIID4hiCDAsoUA4DAAbgeqmSJHOfwtAoDzpHOUB5aIKz0EwygiOJUF4UkED2JgWCFYoyj5UoPAATwfD9rCCAAMJSMIMAoCIYFoDwQA)

Screenshot 2024-08-08 at 10 26 12 AM

## Screenshots
_Note: Optional_

## What areas of the site does it impact?
*(Describe what parts of the site are impacted and*if*code touched other areas)*

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [x]  Events are being sent to the appropriate logging solution
- [x]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x]  Feature/bug has a monitor built into Datadog (if applicable)
- [x]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [x]  I added a screenshot of the developed feature

## Requested Feedback

(OPTIONAL)_What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

### CODE REVIEW

File: chart/templates/hpa-external-metric.yaml [hunk 1]
```diff
@@ -1,8 +1,12 @@
 {{- $root := . -}}
 {{- range $name, $service := $root.Values.web.webServices }}
 {{- if $service.autoscaling.enabled }}
-{{- $pool := "sum:puma.pool_capacity{kube_cluster_name:dsva-vagov-" }}
-{{- $threads := "sum:puma.max_threads{kube_cluster_name:dsva-vagov-" }}
+{{- $pool := "avg:puma.pool_capacity{kube_cluster_name:dsva-vagov-" }}
+{{- $pool_with_filters := "(" $pool $root.Values.target_env "-cluster,kube_deployment:" $name ")" }}
+{{- $threads := "avg:puma.max_threads{kube_cluster_name:dsva-vagov-" }}
+{{- $threads_with_filters := "(" $threads $root.Values.target_env "-cluster,kube_deployment:" $name ")" }}
```
Comment (ericboehs) [inline chart/templates/hpa-external-metric.yaml:7]: What do you think of
```suggestion
{{- $max_threads := "avg:puma.max_threads{kube_cluster_name:dsva-vagov-" }}
{{- $max_available_threads := "(" $max_threads $root.Values.target_env "-cluster,kube_deployment:" $name ")" }}
```

File: chart/templates/hpa-external-metric.yaml [hunk 2]
```diff
@@ -1,8 +1,12 @@
...
+{{- $replicas_available := "sum:kubernetes_state.deployment.replicas_available{kube_cluster_name:dsva-vagov-" }}
+{{- $replicas_with_filters := "(" $replicas_available $root.Values.target_env "-cluster,kube_deployment:" $name ")" }}
```
Comment (ericboehs) [inline chart/templates/hpa-external-metric.yaml:9]: Could these two lines be combined?

File: chart/templates/hpa-external-metric.yaml [hunk 3]
```diff
@@ -12,11 +16,21 @@ metadata:
...
+        ({{ $pool_with_filters }} * {{ $replicas_with_filters }})
+      )
+      /
+      ({{ $threads_with_filters }} * {{ $replicas_with_filters }})
+    )
+    * 100
 {{- end }}
 {{- end }}
 
+
```
Comment (ericboehs) [inline chart/templates/hpa-external-metric.yaml:33]: Extra line break?

File: chart/templates/hpa-external-metric.yaml (hunk 2 above)
Comment (LindseySaari) [reply chart/templates/hpa-external-metric.yaml:9]: I think this is cleaner and easier to understand with them being separate. If we ever decide to change something, it seems more complex if we fold in the first query with the query_with_filters

File: chart/templates/hpa-external-metric.yaml [hunk 4]
```diff
@@ -1,9 +1,13 @@
...
-{{- $pool := "sum:puma.pool_capacity{kube_cluster_name:dsva-vagov-" }}
-{{- $threads := "sum:puma.max_threads{kube_cluster_name:dsva-vagov-" }}
-{{- $deployment := "-cluster,kube_deployment:" }}
+{{- $pool := "avg:puma.pool_capacity{kube_cluster_name:dsva-vagov-" }}
+{{- $threads := "avg:puma.max_threads{kube_cluster_name:dsva-vagov-" }}
+{{- $replicas_available := "sum:kubernetes_state.deployment.replicas_available{kube_cluster_name:dsva-vagov-" }}
+{{- $pool_with_filters := $pool | cat $root.Values.target_env "-cluster,kube_deployment:" $name "}" }}
+{{- $threads_with_filters := $threads | cat $root.Values.target_env "-cluster,kube_deployment:" $name "}" }}
+{{- $replicas_with_filters := $replicas_available | cat $root.Values.target_env "-cluster,kube_deployment:" $name "}" }}
+
```
Comment (rmtolmach) [inline chart/templates/hpa-external-metric.yaml:10]: non-blocking, but there's an extra line here

Comment (LindseySaari) [reply chart/templates/hpa-external-metric.yaml:10]: I put that there to separate the regular vars from the calculations

**Review by rmtolmach — APPROVED** [review]

**Pull request merged**